pkg/controller/podgc/gc_controller.go

```diff
@@ -185,6 +186,114 @@ func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
 	}
 }
 
+type deletedNodesWatcher struct {
+	recentlyDeletedNodes map[string]time.Time
+	existingNodes        sets.String
+	kubeClient           clientset.Interface
+	lastPurge            time.Time
+}
+
+func NewDeletedNodesWatcher(kubeClient clientset.Interface) *deletedNodesWatcher {
+	return &deletedNodesWatcher{
+		recentlyDeletedNodes: make(map[string]time.Time),
+		existingNodes:        sets.NewString(),
+		kubeClient:           kubeClient,
+	}
+}
+
+func (dnw *deletedNodesWatcher) FilterDeletedNames(names sets.String) sets.String {
+	dnw.purgeIfNeeded()
+	deletedNames := sets.NewString()
+	unknownNames := sets.NewString()
+	// Filter against watch data
+	for name := range names {
+		if name == "" {
+			continue
+		}
+		if _, found := dnw.recentlyDeletedNodes[name]; found {
+			deletedNames.Insert(name)
+			continue
+		}
+		if _, found := dnw.existingNodes[name]; found {
+			continue
+		}
+		unknownNames.Insert(name)
+	}
+	if len(unknownNames) > 0 {
```

Comment:
If a user keeps creating pods with fake node names, than updateExistingNodes() will always be called.

I'm thinking if there is a better way to detect if the cached node list is stale.

Reply:
I've changed the implementation. Currently in fake node names scenario, there is a single GET (per ~hour) for every unique node name.

Reply:
Implementation changed once again.
When orphaned pod is found and after 40s there is still no node with matching name in an informer, GET call is issued to verify that the node is truly gone (single call for all pods from the same node, but without caching the result).